### DISCUSSION TITLE
NodeJs后端可以借助编译成动态库的node.dll来实现类似纯V8的代码编写方式吗

### DISCUSSION THREAD

**Discussion by yqs112358** [Tencent/ScriptX#50] [Q&A]
也就是说，将ScriptX后端相关的代码插入在主项目中，而不是作为node的addon加载
node官方仓库文档提供了嵌入式Node的示例 https://github.com/nodejs/node/blob/v14.16.0/doc/api/embedding.md

**Comment by LanderlYoung** [comment]
sure，please checkout the demo

**Reply by yqs112358** [reply]
I have read through the demo carefully, of course.
The demo of node backend is node-addon. 
However, my meaning is to use native node-api like how Electron works, instead of node-addon, as this addon creates a gap between the main code and those interact with node's engine

I think my idea is more similar with this issue https://github.com/nodejs/help/issues/818

It is known that chakra-node is easy for embedding because of its official support.  As the issue said above, node-v8 is not so easy to do so, but considering that those library who use native apis in npm need to be recompiled to adapt to chakra-node, node-v8 is still necessary and irreplaceable.
Without ScriptX's support, chakra-node is difficult to integrate into the cross-language system, too. \

Maybe ScriptX can give a better solution for the community?

**Reply by LanderlYoung** [reply]
Sorry for my misunderstanding.

Currently, ScriptX supports node-addon.

That's because we have support for v8-api, so why not. So as you can see, we support node-addon only be v8-api, not the [napi](https://nodejs.org/dist./v9.11.2/docs/api/n-api.html).
By using ScriptX you can write a node add-on with some code when you embedding v8.

**Reply by yqs112358** [reply]
Sure, I have tested node-addon part of ScriptX and it works well. Undeniable that it's a good choice for those applications based on nodejs (for example, their main code is in Js）
然而，这个方案的问题在于：如果你需要像嵌入V8一样嵌入node到某个C++项目，这会遇到问题。在纯V8方案中，C++项目的主要部分的代码和ScriptX部分操纵引擎的代码最终将编译为同一个可执行模块，最终整个可执行文件只需要链接到v8即可。
但是如果使用node-addon方案的ScriptX，操纵脚本引擎的代码最终将编译为.node的node子模块，这就导致其和C++项目主要部分的代码割裂开，导致很多问题。

对于这个问题，最好的解决方案是将node像V8那样抽象为一个引擎，让他像嵌入式的V8一样作为动态链接库工作
electon用的就是类似这个的方案，他们根据自己的经验做了一些说明：https://www.electronjs.org/blog/electron-internals-using-node-as-a-library

其实，对于ScriptX来说，这个设计只需要在现有v8 backend的基础上做一些小的改动即可。
- 将node编译为动态链接库或者静态链接
- 将引擎初始化部分做一些改动，将单纯的V8::Initialize上加上针对node的初始化（比如支持worker而调用的node::MultiIsolatePlatform::Create）
- 把“创建一个ScriptEngine”抽象为创建一个新的node environment (CommonEnvironmentSetup)，销毁一个ScriptEngine抽象为销毁那个node environment（node::Stop(env) ）
- 可以将node所需要的的libuv事件循环（node::SpinEventLoop）塞进ScriptEngine的MessageLoop
其余绝大部分的逻辑仍然沿用v8-ScriptX，不需要改动。这样，由于有Node基础库和npm的加持，js部分的能力可以说是如虎添翼

在node官方的嵌入式指导文档（[https://github.com/nodejs/node/blob/master/doc/api/embedding.md](https://github.com/nodejs/node/blob/master/doc/api/embedding.md)）中提到了上面几乎所有的东西，这也许可以起到很大的帮助

![image](https://user-images.githubusercontent.com/37969157/137425696-7bfdadea-d687-4f71-a2bc-85cc08ae5248.png)
